daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala

```diff
@@ -9,84 +9,89 @@ import com.daml.ledger.api.domain.PartyDetails
 import com.daml.script.converter.ConverterException
 import io.grpc.StatusRuntimeException
+import org.slf4j.LoggerFactory
 import scalaz.OneAnd
 import scalaz.OneAnd._
 import scalaz.std.set._
 import scalaz.syntax.foldable._
 
 import scala.collection.compat.immutable.LazyList
-import scala.collection.mutable.ArrayBuffer
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
 // Client for the script service.
 class IdeLedgerClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClient {
-  class ArrayBufferTraceLog extends TraceLog {
-    val buffer = ArrayBuffer[(String, Option[Location])]()
-    override def add(message: String, optLocation: Option[Location]): Unit = {
-      discard { buffer.append((message, optLocation)) }
-    }
-    override def iterator: Iterator[(String, Option[Location])] = {
-      buffer.iterator
-    }
-    def clear: Unit = buffer.clear()
-  }
+  private val pkg2TxVersion =
+    compiledPackages.interface.packageLanguageVersion.andThen(
+      TransactionVersion.assignNodeVersion
+    )
+  private val contractKeyUniqueness = ContractKeyUniquenessMode.On
+  private val damlTraceLog = LoggerFactory.getLogger("daml.tracelog")
 
-  val traceLog = new ArrayBufferTraceLog()
+  import IdeLedgerClient.SubmissionCache
 
-  private[this] val preprocessor = new engine.preprocessing.CommandPreprocessor(compiledPackages)
+  private var seed = crypto.Hash.hashPrivateKey(s"script-service")
 
-  private val seed = crypto.Hash.hashPrivateKey(s"script-service")
+  private var _lastSubmission: Option[SubmissionCache] = None
 
-  private val txSeeding =
-    speedy.InitialSeeding.TransactionSeed(seed)
+  private[this] def emptyPtx: PartialTransaction = PartialTransaction.initial(
+    pkg2TxVersion,
+    contractKeyUniqueness,
+    ledger.currentTime,
+    InitialSeeding.TransactionSeed(seed),
+  )
 
-  // Machine for submissions.
-  val machine = Machine(
-    compiledPackages,
-    submissionTime = Time.Timestamp.Epoch,
-    initialSeeding = txSeeding,
-    expr = null,
-    globalCids = Set.empty,
-    committers = Set.empty,
-    traceLog = traceLog,
+  private[this] def clearPtx(): Unit =
+    _lastSubmission = _lastSubmission.map(cache =>
+      cache.copy(
+        ptx = emptyPtx
+      )
+    )
+
+  def lastSubmission: SubmissionCache = _lastSubmission.getOrElse(
+    SubmissionCache(
+      ptx = emptyPtx,
+      traceLog = RingBufferTraceLog(damlTraceLog, 100),
+      commitLocation = None,
+    )
   )
-  val onLedger = machine.ledgerMode match {
-    case OffLedger => throw SRequiresOnLedger("ScenarioRunner")
-    case onLedger: OnLedger => onLedger
-  }
-  val scenarioRunner = ScenarioRunner(machine, seed)
+
+  private[this] val preprocessor = new engine.preprocessing.CommandPreprocessor(compiledPackages)
+
+  private var _ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
+  def ledger: ScenarioLedger = _ledger
```

Comment:
:trollface: 
```suggestion
  private var ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
  def getLedger: ScenarioLedger = ledger
```
Just trolling, please don't do this.

Reply:
working on adding a lombok dependency to autogneerate this